syntax = "proto3";

message ZexState {
    map<string, Market> markets = 1;
    map<string, Balance> balances = 2;
    repeated AmountEntry amounts = 3;
    repeated TradeEntry trades = 4;
    repeated OrderEntry orders = 5;
    map<string, WithdrawsOnChain> withdraws_on_chain = 6;
    map<string, UserWithdraws> user_withdraws_on_chain = 7;
    map<string, WithdrawNonceOnChain> user_withdraw_nonce_on_chain = 8;
    map<string, uint64> withdraw_nonce_on_chain = 9;
    
    map<string, DepositsOnChain> deposits = 10;
    repeated NonceEntry nonces = 11;
    uint64 last_tx_index = 12;
    repeated UserDepositEntry user_deposits = 13;
    repeated IDLookupEntry public_to_id_lookup = 14;
    map<uint64, bytes> id_to_public_lookup = 15;
    map<string, ContractDecimalOnChain> contract_decimal_on_chain = 16;
}

message Market {
    string base_token = 1;
    string quote_token = 2;
    repeated Order buy_orders = 3;
    repeated Order sell_orders = 4;
    repeated OrderBookEntry bids_order_book = 5;
    repeated OrderBookEntry asks_order_book = 6;
    uint64 first_id = 7;
    uint64 final_id = 8;
    uint64 last_update_id = 9;
    bytes kline = 10; // Serialized pandas DataFrame
}

message Order {
    string price = 1;
    bytes tx = 2;
}

message OrderBookEntry {
    string price = 1;
    string amount = 2;
}

message Balance {
    repeated BalanceEntry balances = 1;
}

message BalanceEntry {
    bytes public_key = 1;
    string amount = 2;
}

message AmountEntry {
    bytes tx = 1;
    string amount = 2;
}

message TradeEntry {
    bytes public_key = 1;
    repeated Trade trades = 2;
}

message Trade {
    uint32 t = 1;
    string amount = 2;
    string pair = 3;
    uint32 order_type = 4;
    bytes order = 5;
}

message OrderEntry {
    bytes public_key = 1;
    repeated bytes orders = 2;
}

message UserWithdrawEntry {
    bytes public_key = 1;
    repeated bytes raw_txs = 2;
}

message UserWithdraws {
    repeated UserWithdrawEntry withdraws = 1;
}

message WithdrawsOnChainEntry {
    repeated bytes raw_txs = 1;
}
message WithdrawsOnChain {
    repeated WithdrawsOnChainEntry withdraws = 1;
}

message WithdrawNonceEntry {
    bytes public_key = 1;
    uint64 nonce = 2;
}

message WithdrawNonceOnChain {
    repeated WithdrawNonceEntry nonces = 1;
}

message DepositEntry {
    string tx_hash = 1;
    uint32 vout = 2;
}

message DepositsOnChain {
    repeated DepositEntry deposits = 1;
}

message UserDepositEntry {
    bytes public_key = 1;
    repeated Deposit deposits = 2;
}

message Deposit {
    string tx_hash = 1;
    string chain = 2;
    string token_contract = 3;
    string amount = 4;
    uint32 decimal = 5;
    uint64 time = 6;
    uint64 user_id = 7;
    uint32 vout = 8;
}

message NonceEntry {
    bytes public_key = 1;
    uint32 nonce = 2;
}

message IDLookupEntry {
    bytes public_key = 1;
    uint64 user_id = 2;
}

message ContractToIDOnChainEntry {
    string chain = 1;
    map<string, uint64> contract_to_id = 2;
}

message IDToContractOnChainEntry {
    string chain = 1;
    map<uint64, string> id_to_contract = 2;
}

message ContractDecimalOnChain {
    map<string, uint32> contract_decimal = 1;
}